### PR TITLE
Separate environments and upgraded submission requirements

### PR DESCRIPTION
**Pull request by lacava**
This PR introduces separate environments for each method in SRBench, much like what was done in the 2022 competition. This wil l make it much easier to test and maintain new methods as the project evolves. 

Method submission are now structured to follow the competition 2022 format, which includes requiring methods to output sympy compatible models. 

Methods are contained within folders in the `algorithms` directory (e.g., `algorithms/feat`). This folder should contain:

  1. `metadata.yml` (**required**): A file describing your submission, following the descriptions in [submission/feat-example/metadata.yml][metadata]. 
  2. `regressor.py` (**required**): a Python file that defines your method, named appropriately. See [submission/feat-example/regressor.py][regressor] for complete documentation. 
      It should contain:
      -   `est`: a sklearn-compatible `Regressor` object. 
      -   `model(est, X=None)`: a function that returns a [**sympy-compatible**](https://www.sympy.org) string specifying the final model. It can optionally take the training data as an input argument. See [guidance below](###-returning-a-sympy-compatible-model-string). 
      -   `eval_kwargs` (optional): a dictionary that can specify method-specific arguments to `evaluate_model.py`.
  3. `LICENSE` *(optional)* A license file
  4. `environment.yml` *(optional)*: a [conda environment file](https://docs.conda.io/projects/conda/en/latest/user-guide/tasks/manage-environments.html#creating-an-environment-from-an-environment-yml-file) that specifies dependencies for your submission. 
  It will be used to update the baseline environment (`environment.yml` in the root directory). 
  To the extent possible, conda should be used to specify the dependencies you need. 
  If your method is part of conda, great! You can just put that in here and leave `install.sh` blank. 
  5. `requirements.txt` (option): a pypi requirements file. The script will run `pip install -r requirements.txt` if this file is found, before proceeding.
  5. `install.sh` *(optional)*: a bash script that installs your method. 
  **Note: scripts should not require sudo permissions. The library and include paths should be directed to conda environment; the environmental variable `$CONDA_PREFIX` specifies the path to the environment.
  6. additional files *(optional)*: you may include a folder containing the code for your method in the submission. However it is much preferred to pull your code from a repo within `install.sh`. 

# TODO 

- [ ] Since the requirements have been updated and the tests have been rewritten, at the moment, many methods are not passing the tests. Will merge once all methods are passing.
- [ ] updated docs on submitting algorithms. 

## Algorithm status

**If you're tagged below, please take a look and see if you can help get your algorithm to install and test correctly. Thank you!**

- [x] afp  
- [x] geneticengine 
- [x] operon @foolnotion 
- [x] bingo 
- [x] gpgomea @marcovirgolin
- [x] ps-tree @hengzhe-zhang 
- [ ] dsr    @brendenpetersen 
- [x] gplearn        
- [x] e2et   
- [x] gpzgd          
- [x] pysr @MilesCranmer (maybe just needs the test params to be set?)
- [x] eplex  
- [ ] gsgp           
- [x] qlattice
- [x] eql    
- [x] itea           
- [ ] sbp-gp
- [x] feat   
- [ ] mrgp           
- [ ] taylorgp
- [x] ffx    
- [ ] nsga-dcgp      
- [ ] uDSR @brendenpetersen 
- [x] TIR @folivetti 
- [ ] DAG search

### CODE REVIEW

**Comment by lacava** [comment]
this PR will also fix #58

**Comment by MilesCranmer** [comment]
Thanks. Could you set the test params for PySR to the following?

```python
eval_kwargs = {
    "test_params": dict(
        niterations=3,
        ncyclesperiteration=500,
        population_size=3,
    )
}
```

that should be enough to fix things.

**Comment by folivetti** [comment]
you can also check

[x] TIR

:-) 

I'll have a look at the other methods if there's something I can quickly fix

**Comment by MilesCranmer** [comment]
@folivetti could you add my suggestions for fixing PySR? I don't have push permissions unfortunately

**Comment by folivetti** [comment]
> @folivetti could you add my suggestions for fixing PySR? I don't have push permissions unfortunately

certainly. Should I put it in this file? https://github.com/cavalab/srbench/blob/separate-envs/algorithms/pysr/regressor.py

**Comment by lacava** [comment]
@folivetti @MilesCranmer i just added them a few minutes ago

**Comment by lacava** [comment]
upgraded your permissions @MilesCranmer

**Comment by MilesCranmer** [comment]
Ping on this...

**Comment by lacava** [comment]
trying to find some help to debug the remaining 13 methods building in the new setup with separate environments. please help if you can! 
@hengzhe-zhang @foolnotion @marcovirgolin @landajuela

**Comment by MilesCranmer** [comment]
@lacava Since everything is separate now, there's no longer a need to wait for everything to be working at once (which might be too ambitious).

Master branch has been broken for a while already, whereas with this PR, 70% of the algorithms are working. So maybe just merge now, and people can push fixes for individual algorithms at their own pace?

**Comment by folivetti** [comment]
> @lacava Since everything is separate now, there's no longer a need to wait for everything to be working at once (which might be too ambitious).
> 
> Master branch has been broken for a while already, whereas with this PR, 70% of the algorithms are working. So maybe just merge now, and people can push fixes for individual algorithms at their own pace?

following this comment, we can create a separate-envs-broken branch only with the current broken algorithms that will eventually get merged to master again.

What are the current algorithm forked and maintained by you? GPSP and FFX?

**Comment by hengzhe-zhang** [comment]
I have fixed some bugs in my code to make it compatible with the new DEAP package. Now, just need to trigger the check again.

**Comment by lacava** [comment]
> @lacava Since everything is separate now, there's no longer a need to wait for everything to be working at once (which might be too ambitious).
> 
> Master branch has been broken for a while already, whereas with this PR, 70% of the algorithms are working. So maybe just merge now, and people can push fixes for individual algorithms at their own pace?

I see your point, I think we can merge without everything fixed. But while master branch is failing this is really due to only 1 maybe 2 algs not passing. So I'd like to get a few more working instead of breaking ones that currently work. 

> following this comment, we can create a separate-envs-broken branch only with the current broken algorithms that will eventually get merged to master again.

i like this idea @folivetti 

> What are the current algorithm forked and maintained by you? GPSP and FFX?

DSR is also a fork at the moment. MRGP as well.

**Comment by folivetti** [comment]
I'll have a look at those later. FFX seems easy to fix.

**Comment by lacava** [comment]
@marcovirgolin put the furniture instructions down and save your baby

**Comment by lacava** [comment]
thanks @foolnotion for the updates!

**Comment by lacava** [comment]
> Is there anything I can help with? I keep seeing SR papers that exclude PySR as one of the baseline models simply because it's not in SRBench 🙁 ... Would be great to get this all working

i'll plan to get it merged this week.

**Comment by lacava** [comment]
for some reason pysr is failing again @MilesCranmer

**Comment by MilesCranmer** [comment]
My guess is that the base conda env is not fixing all the versions, and was just updated to Python 3.12, which e.g., the conda-forge version of PySR was not yet packaged for.

```
Could not solve for environment specs
The following packages are incompatible
└─ pysr 0.16  is installable with the potential options
   ├─ pysr 0.16.0 would require
   │  └─ python >=3.10,<3.11.0a0 , which can be installed;
   ├─ pysr 0.16.0 would require
   │  └─ python >=3.11,<3.12.0a0 , which can be installed;
   ├─ pysr 0.16.0 would require
   │  └─ python >=3.8,<3.9.0a0 , which can be installed;
   └─ pysr 0.16.0 would require
      └─ python >=3.9,<3.10.0a0 , which can be installed.
```

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -51,137 +37,161 @@ jobs:
           activate-environment: srbench
           use-mamba: true
```
Comment (MilesCranmer) [inline .github/workflows/ci.yml:38]:
```suggestion
          use-mamba: true
          python-version: 3.11
```

Looks like the Python version was not set here, so recently updated to 3.12 which will break a bunch of packages that haven't yet released their 3.12 versions (PyTorch usually takes a while to do so, for example).

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -51,137 +37,161 @@ jobs:
...
+              - format('algorithms/{0}/**', matrix.alg)
       - 
         name: Setup Mambaforge
+        # if: steps.changes.outputs.alg == 'true'
         uses: conda-incubator/setup-miniconda@v2
         with:
           miniforge-variant: Mambaforge
           miniforge-version: latest
           activate-environment: srbench
           use-mamba: true
```
Comment (MilesCranmer) [inline .github/workflows/ci.yml:112]:
```suggestion
          use-mamba: true
          python-version: 3.11
```

File: algorithms/pysr/environment.yml [hunk 3]
```diff
@@ -0,0 +1,4 @@
+channels:
+    - conda-forge
+dependencies:
+    - pysr==0.16
```
Comment (MilesCranmer) [inline algorithms/pysr/environment.yml:4]:
```suggestion
    - pysr==0.16
    - python==3.11
```

Comment (MilesCranmer) [reply algorithms/pysr/environment.yml:4]: (This will also fix PySR, but it's safer to just set Python 3.11 globally)

Comment (lacava) [reply algorithms/pysr/environment.yml:4]: Some of the algs need other versions of python, so maybe it's best to set it in pysr

**Pull request merged**